server/tools.go

```diff
@@ -0,0 +1,175 @@
+				continue
+			case errors.As(err, &unmarshalType):
+				offset += int(unmarshalType.Offset)
+				continue
+			default:
+				return nil
+			}
+		}
+		offset += int(decoder.InputOffset())
+		objs = append(objs, obj)
+	}
+	return objs
+}
+
+// parseJSONToolCalls attempts to parse a JSON string into a slice of ToolCalls.
+// Returns parsed tool calls and a boolean indicating if the JSON is incomplete
+func parseJSONToolCalls(tmpl *gotmpl.Template, s string) ([]api.ToolCall, bool) {
+	var b bytes.Buffer
+	if err := tmpl.Execute(&b, map[string][]api.ToolCall{
+		"ToolCalls": {
+			{
+				Function: api.ToolCallFunction{
+					Name: "@@name@@",
+					Arguments: api.ToolCallFunctionArguments{
+						"@@argument@@": 1,
+					},
+				},
+			},
+		},
+	}); err != nil {
+		return nil, false
+	}
+
+	templateObjects := parseObjects(b.String())
+	if len(templateObjects) == 0 {
+		return nil, false
+	}
+
+	// find the keys that correspond to the name and arguments fields
+	var name, arguments string
+	for k, v := range templateObjects[0] {
+		switch v.(type) {
+		case string:
+			name = k
+		case map[string]any:
+			arguments = k
+		}
+	}
+
+	if name == "" || arguments == "" {
+		return nil, false
+	}
+
+	responseObjects := parseObjects(s)
+	if len(responseObjects) == 0 {
+		return nil, false
+	}
+
+	// collect all nested objects
+	var collect func(any) []map[string]any
+	collect = func(obj any) (all []map[string]any) {
+		switch o := obj.(type) {
+		case map[string]any:
+			all = append(all, o)
+			for _, v := range o {
+				all = append(all, collect(v)...)
+			}
+		case []any:
+			for _, v := range o {
+				all = append(all, collect(v)...)
+			}
+		}
+
+		return all
+	}
+
+	var objs []map[string]any
+	for _, p := range responseObjects {
+		objs = append(objs, collect(p)...)
+	}
+
+	var toolCalls []api.ToolCall
+	for _, kv := range objs {
+		n, nok := kv[name].(string)
+		a, aok := kv[arguments].(map[string]any)
+		if nok && aok {
+			toolCalls = append(toolCalls, api.ToolCall{
+				Function: api.ToolCallFunction{
+					Name:      n,
+					Arguments: a,
+				},
+			})
+		}
+	}
+
+	return toolCalls, len(toolCalls) > 0
+}
+
+// routeToolParsing is a helper function that routes what kind of tool parsing to use
+func routeToolParsing(s string, tmpl *gotmpl.Template) ([]api.ToolCall, bool, bool) {
+	if strings.HasPrefix(s, "[{") || strings.HasPrefix(s, "```") || strings.HasPrefix(s, "{") {
+		if toolCalls, ok := parseJSONToolCalls(tmpl, s); ok {
+			return toolCalls, false, true
+		}
+		// in the case the JSON never finishes, the acuumulated content should be sent downstream
+		return nil, true, true
+	}
+	// TODO(parthsareen): add python tool call support
+	return nil, false, false
+}
+
+// ParseToolCalls extracts tool calls from a string using a tool token prefix or direct JSON parsing.
+// Returns tool calls, whether parsing is incomplete, and any errors.
+func ParseToolCalls(s string, toolToken string, tmpl *gotmpl.Template) ([]api.ToolCall, bool, error) {
```

Comment:
As separate suggestion, we might want to bundle some state and functions into a struct type here that can read the next token and determine whether a tool was called, whether it can never be called (e.g. wrong prefix) or whether to wait, etc

Reply:
Here's an idea of what that could look like:

```
type toolParser struct {
  sb strings.Builder
}

func newToolParser(template.Template) (toolParser, error) {
  // find the prefix
  // get the tool template
  // initialize state
}

// accept will take a response or part of a response and attempt to parse tools  based on what its seen so far, returning true if a partial tool has been found
func (tp *toolParser) accept() ([]api.ToolCall, bool, error) {
  // parse prefix
  // once prefix is parsed, accumulate and parse tools, dismissing non tool tokens like "," or anything else
}
```